android/java/org/chromium/chrome/browser/app/domain/KeyringModel.java

```diff
@@ -264,38 +279,59 @@ public void getAccounts(Callbacks.Callback1<AccountInfo[]> callback1) {
-                    mKeyringService.getKeyringInfo(
-                            getSelectedCoinKeyringId(coinType), updatedKeyringInfo -> {
-                                for (AccountInfo accountInfo : updatedKeyringInfo.accountInfos) {
-                                    if (accountInfo.coin == coinType) {
-                                        setSelectedAccount(accountInfo.address, coinType);
-                                        break;
-                                    }
-                                }
-                            });
-                }
-            }
-            mCryptoSharedActions.updateCoinType();
-            mCryptoSharedActions.onNewAccountAdded();
-            callback.call(result);
-        });
+        if (coinType == CoinType.FIL) {
+            mKeyringService.addFilecoinAccount(
+                    accountName, BraveWalletConstants.FILECOIN_MAINNET, result -> {
+                        handleAccountResult(result, finalAccountInfo, coinType, callback);
+                    });
+        } else {
+            mKeyringService.addAccount(accountName, coinType, result -> {
+                handleAccountResult(result, finalAccountInfo, coinType, callback);
+            });
+        }
     }
 
     public void isWalletLocked(Callbacks.Callback1<Boolean> callback) {
         mKeyringService.isLocked(isWalletLocked -> callback.call(isWalletLocked));
     }
 
+    private void handleAccountResult(boolean result, AccountInfo[] finalAccountInfo,
+            @CoinType.EnumType int coinType, Callbacks.Callback1<Boolean> callback) {
+        if (result) {
+            boolean hasExistingAccountType = false;
+            for (AccountInfo accountInfo : finalAccountInfo) {
+                hasExistingAccountType = (accountInfo.coin == coinType);
+                if (hasExistingAccountType) break;
+            }
+            if (!hasExistingAccountType) {
+                mKeyringService.getKeyringInfo(
+                        getSelectedCoinKeyringId(coinType), updatedKeyringInfo -> {
+                            for (AccountInfo accountInfo : updatedKeyringInfo.accountInfos) {
+                                if (accountInfo.coin == coinType) {
+                                    setSelectedAccount(accountInfo.address, coinType);
+                                    break;
+                                }
+                            }
+                        });
+            }
+        }
+        mCryptoSharedActions.updateCoinType();
+        mCryptoSharedActions.onNewAccountAdded();
+        callback.call(result);
+    }
+
+    private void handleSelectedAccount(String accountAddress) {
+        if (accountAddress == null) {
+            mKeyringService.getKeyringInfo(getSelectedCoinKeyringId(mSharedData.getCoinType()),
+                    keyringInfo -> { updateSelectedAccountPerOriginOrFirst(keyringInfo); });
+        } else {
+            update();
+        }
```

Comment:
The logic inside this method is untouched and was simply copy-pasted in order to be used as a response to `addFilecoinAccount`/`addAccount`.  
There's maybe some room for improvement here? I haven't investigated, but when `result` is false it feels logical to me skipping `updateCoinType()`, `onNewAccountAdded()`.

Reply:
I think this is regarding `handleAccountResult`.
The selected account is only updated after positive [`addAccount` result](https://github.com/brave/brave-core/blob/master/android/java/org/chromium/chrome/browser/app/domain/KeyringModel.java#L271) so update the new implementation accodingly.

Reply:
Github automatically posted this as a separate review [comment](https://github.com/brave/brave-core/pull/18144#discussion_r1171218215) (not the first time). Deleted the duplicate.

Reply:
> The selected account [is only updated after successful retult](https://github.com/brave/brave-core/blob/master/android/java/org/chromium/chrome/browser/app/domain/KeyringModel.java?rgh-link-date=2023-04-19T11%3A48%3A06Z#L271) add account result so please do the same in the [updated addAccoun](https://github.com/brave/brave-core/blob/29caef874365629a6be43e156c27ae77d8fdcdb0/android/java/org/chromium/chrome/browser/app/domain/KeyringModel.java#L286-L293)t.

I'm referring to these two lines:

```java
        mCryptoSharedActions.updateCoinType();
        mCryptoSharedActions.onNewAccountAdded();
```

These are called _always_ regardless of the outcome of `result`. Is this correct?